File: src/commands/d3mantle.ts
import Demantle from '../demantle/Demantle';
import { ErrorEmbed } from '../core/functions.js';
import {
    ActivityType, EmbedBuilder,
    SlashCommandBuilder, PresenceUpdateStatus,
    type ChatInputCommandInteraction, type CacheType,
} from 'discord.js';

import type { Command } from '../typings/core';
import type { Bot } from '../core/client';

const getGameOrError = (client: Bot, channelId: string) => {
    const demantle = client.demantles.get(channelId);
    if (!demantle)
        throw ErrorEmbed("D3mantle Error", "There is no game in progress in this channel.");

    return demantle;
};

const startNewGame = async (interaction: ChatInputCommandInteraction<CacheType>, gameType: 'Random' | 'Daily') => {
    if (interaction.client.demantles.has(interaction.channel!.id))
        throw ErrorEmbed("D3mantle Error", "A game is already in progress in this channel.");

    interaction.client.demantles.set(
        interaction.channel!.id,
        { game: new Demantle(gameType), ignoreIds: [], message: null }
    );

    // Update bot presence to show it's playing d3mantle
    interaction.client.user.setPresence({
        activities: [{
            name: 'd3mantle',
            state: gameType === 'Random' ? "Random Word" : "Daily Word",
            type: ActivityType.Playing,
        }],
        status: PresenceUpdateStatus.Online,
    });

    const title = gameType === 'Random' ? "D3mantle: Random Word" : "D3mantle: Daily Word";
    await interaction.editReply({
        embeds: [
            new EmbedBuilder()
                .setTitle(title)
                .setDescription("Game initiated, you can start guessing already!")
                .setColor("Gold")
        ]
    });
};

export default <Command>{
    data: new SlashCommandBuilder()
        .setName("d3mantle")
        .setDescription("Play a game of d3mantle which is a discord port of Semantle")
        .addSubcommand(subCommand =>
            subCommand
                .setName("random")
                .setDescription("Start a game with a random word")
        )
        .addSubcommand(subCommand =>
            subCommand
                .setName("daily")
                .setDescription("Start a game with the daily word")
        )
        .addSubcommand(subCommand =>
            subCommand
                .setName("giveup")
                .setDescription("Give up on the current game")
                .addBooleanOption(option =>
                    option
                        .setName("show_publicly")
                        .setDescription("Do you want to show the answer publicly?")
                        .setRequired(true)
                )
        )
        .addSubcommand(subCommand =>
            subCommand
                .setName("board")
                .setDescription("Show the game board again if it got buried by chat")
        ),
    async execute(interaction) {
        await interaction.deferReply();

        if (!interaction.channel)
            return await interaction.editReply({
                content: "This command can only be used in a text channel.",
            });

        if (!interaction.guild || interaction.channel.isDMBased())
            return await interaction.editReply({
                content: "This command is not designed to be used in DMs.",
            });

        if (interaction.channel.name !== 'd3mantle')
            throw ErrorEmbed(
                "Channel Restriction",
                "You can only run this command in the `d3mantle` channel due to spamming nature of the game. If the channel doesn't exist, please ask a server admin to create the channel and make sure the channel name is exactly `d3mantle`. You may optionally setup a role system to give access to the players to avoid notification spam for those who are not interested in playing."
            );

        const subCommand = interaction.options.getSubcommand(true) as 'random' | 'daily' | 'giveup' | 'board';

        if (subCommand === 'random')
            await startNewGame(interaction, 'Random');

        else if (subCommand === 'daily')
            await startNewGame(interaction, 'Daily');

        else if (subCommand === 'giveup') {
            const demantle = getGameOrError(interaction.client, interaction.channel.id);
            const isPublicForfeit = interaction.options.getBoolean("show_publicly", true);
            const word = demantle.game.word;

            if (isPublicForfeit) {
                interaction.client.demantles.delete(interaction.channel.id);

                // Clear bot presence if no games are active
                if (interaction.client.demantles.size === 0)
                    interaction.client.user.setPresence({
                        activities: [],
                        status: PresenceUpdateStatus.Idle,
                    });

                await interaction.editReply({
                    embeds: [
                        new EmbedBuilder()
                            .setTitle("D3mantle: Give Up")
                            .setDescription(`No one guessed the word. It was: ${word}`)
                            .setColor("White")
                    ]
                });
            }

            else {
                demantle.ignoreIds.push(interaction.user.id);
                await interaction.editReply({
                    embeds: [
                        new EmbedBuilder()
                            .setTitle("D3mantle: Give Up")
                            .setDescription(`You have given up the game, your guesses will be ignored onwards. The word was: ${word}`)
                            .setColor("White")
                    ]
                });
            }
        }

        else if (subCommand === 'board') {
            const demantle = getGameOrError(interaction.client, interaction.channel.id);

            if (!demantle.message)
                throw ErrorEmbed("D3mantle Error", "No game board exists yet. Make your first guess to generate the board!");

            await interaction.editReply({ content: "Sure, here's the board again" });

            const oldMsg = await demantle.message.delete();
            demantle.message = await interaction.channel.send(oldMsg.content);
        }
    },
}

File: Commands/d3mantle.py
from Demantle.game import Demantle
from disnake.ext import commands
from disnake import (
    Activity,
    ActivityType,
    ApplicationCommandInteraction,
    Colour,
    Embed,
    Option,
    OptionType,
    Status,
    TextChannel,
)

class D3mantleCog(commands.Cog):
    def __init__(self, bot: commands.InteractionBot):
        self.bot = bot

    async def validation(self, interaction: ApplicationCommandInteraction):
        if not interaction.channel:
            await interaction.edit_original_response(content="This command can only be used in a text channel.")
            return False

        if not interaction.guild or not isinstance(interaction.channel, TextChannel):
            await interaction.edit_original_response(content="This command is designed to be used in Server text channels only.")
            return False

        if interaction.channel.name != "d3mantle":
            await interaction.edit_original_response(
                embed=Embed(
                    title="Channel Restriction",
                    description="You can only run this command in the `d3mantle` channel due to spamming nature of the game. "
                                "If the channel doesn't exist, please ask a server admin to create one named exactly `d3mantle`.",
                    colour=Colour.red()
                )
            )
            return False
        return True

    async def start_new_game(self, interaction: ApplicationCommandInteraction, game_type: str):
        channel = interaction.channel.id

        if channel in self.bot.d3mantles: # type: ignore
            await interaction.edit_original_response(
                embed=Embed(
                    title="D3mantle Error",
                    description="A game is already in progress in this channel.",
                    colour=Colour.red()
                )
            )
            return

        self.bot.d3mantles[channel] = { # type: ignore
            "game": Demantle(game_type),
            "ignore_ids": [],
            "message": None,
        }

        await self.bot.change_presence(
            status=Status.online,
            activity=Activity(
                name='d3mantle',
                state="Random word" if game_type == "random" else "Daily word",
                type=ActivityType.playing
            )
        )

        await interaction.edit_original_response(
            embed=Embed(
                title=f"D3mantle: {game_type.capitalize()} Word",
                description="Game initiated! You can start guessing already!",
                colour=Colour.gold()
            )
        )

    @commands.slash_command(
        name="d3mantle",
        description="Play a game of D3mantle (Discord port of Semantle)"
    )
    async def d3mantle(self, interaction: ApplicationCommandInteraction):
        # just a base group command; subcommands below
        pass

    @d3mantle.sub_command(name="random", description="Start a game with a random word")
    async def random_game(self, interaction: ApplicationCommandInteraction):
        await interaction.response.defer()
        if await self.validation(interaction):
            await self.start_new_game(interaction, "random")

    @d3mantle.sub_command(name="daily", description="Start a game with the daily word")
    async def daily_game(self, interaction: ApplicationCommandInteraction):
        await interaction.response.defer()
        if await self.validation(interaction):
            await self.start_new_game(interaction, "daily")

    @d3mantle.sub_command(
        name="give_up",
        description="Give up on the current game",
        options=[
            Option(
                name="show_publicly",
                description="Do you want to show the answer publicly?",
                required=True,
                type=OptionType.boolean,
            )
        ]
    )
    async def give_up(self, interaction: ApplicationCommandInteraction, show_publicly: bool):
        await interaction.response.defer()
        demantle = self.bot.d3mantles.get(interaction.channel.id) # type: ignore

        if demantle is None:
            await interaction.edit_original_response(
                embed=Embed(
                    title="D3mantle Error",
                    description="There is no game in progress in this channel.",
                    colour=Colour.red(),
                )
            )
            return

        word = demantle["game"].word

        if show_publicly:
            del self.bot.d3mantles[interaction.channel.id] # type: ignore
            if not self.bot.d3mantles: # type: ignore
                await self.bot.change_presence(status=Status.idle, activity=None)

            await interaction.edit_original_response(
                embed=Embed(
                    title="D3mantle: Give Up",
                    description=f"No one guessed the word. It was: **{word}**",
                    colour=Colour.from_rgb(255, 255, 255)
                )
            )

        else:
            demantle["ignore_ids"].append(interaction.user.id)
            await interaction.edit_original_response(
                embed=Embed(
                    title="D3mantle: Give Up",
                    description=f"You've given up the game. Your guesses will be ignored onwards.\n"
                                f"The word was `{word}`",
                    colour=Colour.from_rgb(255, 255, 255)
                )
            )

    @d3mantle.sub_command(name="show_board", description="Show the game board again if it got buried by chat")
    async def show_board(self, interaction: ApplicationCommandInteraction):
        await interaction.response.defer()
        demantle = self.bot.d3mantles.get(interaction.channel.id) # type: ignore

        if demantle is None:
            await interaction.edit_original_response(
                embed=Embed(
                    title="D3mantle Error",
                    description="There is no game in progress in this channel.",
                    colour=Colour.red(),
                )
            )
            return

        message = demantle["message"]

        if message is None:
            await interaction.edit_original_response(
                embed=Embed(
                    title="D3mantle Error",
                    description="No game board exists yet. Make your first guess to generate the board!",
                    colour=Colour.red(),
                )
            )
            return

        demantle["message"] = await interaction.edit_original_response(content=f"Sure, here's the board again:\n{message.content}")
        await message.delete()


def setup(bot: commands.InteractionBot):
    bot.add_cog(D3mantleCog(bot))
